File: modules/expo-radial-chart/android/src/main/java/expo/modules/radialchart/ExpoRadialChartView.kt
package expo.modules.radialchart

import android.content.Context
import android.graphics.Color
import androidx.annotation.ColorInt
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.highlight.Highlight
import com.github.mikephil.charting.listener.OnChartValueSelectedListener
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.views.ExpoView


class Series : Record {
  @Field
  val color: String = "#ff0000"

  @Field
  val percentage: Float = 0.0f
}

class ExpoRadialChartView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  private var seriesData: ArrayList<Series> = ArrayList()
  
  internal val chartView = PieChart(context).also {
    it.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    
    // Configure chart appearance
    it.description.isEnabled = false
    it.isDrawHoleEnabled = true
    it.setHoleColor(Color.TRANSPARENT)
    it.holeRadius = 40f
    it.transparentCircleRadius = 45f
    it.setDrawCenterText(false)
    it.legend.isEnabled = false
    it.setTouchEnabled(true) // Enable touch interactions
    it.isRotationEnabled = false
    it.isHighlightPerTapEnabled = true // Enable highlighting on tap
    
    // Set up value selection listener
    it.setOnChartValueSelectedListener(object : OnChartValueSelectedListener {
      override fun onValueSelected(e: com.github.mikephil.charting.data.Entry?, h: Highlight?) {
        e?.let { entry ->
          h?.let { highlight ->
            val index = highlight.x.toInt()
            if (index < seriesData.size) {
              val series = seriesData[index]
              // Send event to React Native
              sendEvent("onSegmentSelect", mapOf(
                "index" to index,
                "percentage" to series.percentage,
                "color" to series.color
              ))
            }
          }
        }
      }
      
      override fun onNothingSelected() {
        // Optional: handle when nothing is selected
      }
    })
    
    addView(it)
  }

  fun setChartData(data: ArrayList<Series>) {
    seriesData = data // Store for event handling
    val entries: ArrayList<PieEntry> = ArrayList()
    val colors: ArrayList<Int> = ArrayList()
    for (series in data) {
      entries.add(PieEntry(series.percentage))
      colors.add(Color.parseColor(series.color))
    }
    val dataSet = PieDataSet(entries, "DataSet")
    dataSet.colors = colors
    dataSet.setDrawValues(false) // Hide percentage labels on slices
    dataSet.sliceSpace = 2f // Add space between slices
    dataSet.selectionShift = 8f // Add offset when selected for hover effect
    
    val pieData = PieData(dataSet)
    chartView.data = pieData
    chartView.invalidate()
  }
}

File: modules/expo-radial-chart/ios/ExpoRadialChartView.swift
import ExpoModulesCore
import DGCharts

struct Series: Record {
  @Field
  var color: UIColor = UIColor.black

  @Field
  var percentage: Double = 0
}

class ExpoRadialChartView: ExpoView {
  let chartView = PieChartView()
  private var seriesData: [Series] = []
  
  // Event dispatcher for segment selection
  let onSegmentSelect = EventDispatcher()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true
    
    // Configure chart appearance
    chartView.legend.enabled = false
    chartView.chartDescription.enabled = false
    chartView.holeRadiusPercent = 0.4
    chartView.transparentCircleRadiusPercent = 0.45
    chartView.drawHoleEnabled = true
    chartView.holeColor = UIColor.clear
    chartView.isUserInteractionEnabled = true // Enable user interaction
    chartView.rotationEnabled = false
    chartView.highlightPerTapEnabled = true // Enable highlighting on tap
    
    // Set up delegate for handling selections
    chartView.delegate = self
    
    addSubview(chartView)
  }

  override func layoutSubviews() {
    chartView.frame = bounds
  }

  func setChartData(data: [Series]) {
    seriesData = data // Store for event handling
    let set1 = PieChartDataSet(entries: data.map({ (series: Series) -> PieChartDataEntry in
      return PieChartDataEntry(value: series.percentage)
    }))
    set1.colors = data.map({ (series: Series) -> UIColor in
      return series.color
    })
    set1.drawValuesEnabled = false // Hide percentage labels on slices
    set1.sliceSpace = 2.0 // Add space between slices
    set1.selectionShift = 8.0 // Add offset when selected for hover effect
    
    let chartData: PieChartData = [set1]
    chartView.data = chartData
  }
}

// MARK: - ChartViewDelegate
extension ExpoRadialChartView: ChartViewDelegate {
  func chartValueSelected(_ chartView: ChartViewBase, entry: ChartDataEntry, highlight: Highlight) {
    let index = Int(highlight.x)
    if index < seriesData.count {
      let series = seriesData[index]
      // Convert UIColor back to hex string
      var red: CGFloat = 0
      var green: CGFloat = 0
      var blue: CGFloat = 0
      var alpha: CGFloat = 0
      series.color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
      let hexColor = String(format: "#%02X%02X%02X", 
                           Int(red * 255), 
                           Int(green * 255), 
                           Int(blue * 255))
      
      // Send event to React Native using EventDispatcher
      onSegmentSelect([
        "index": index,
        "percentage": series.percentage,
        "color": hexColor
      ])
    }
  }
  
  func chartValueNothingSelected(_ chartView: ChartViewBase) {
    // Optional: handle when nothing is selected
  }
}
